File: client/src/store/instance.ts
import { signal } from '@preact/signals'
import { api } from '@/api'

/**
 * Instance metadata fetched from `GET /api/instance/config`.
 *
 * Public endpoint (no token required). The SPA queries this on cold
 * start to decide whether to redirect to `/setup` and which login flow
 * to render. Re-fetched only on cold start; once the SPA has any
 * evidence the instance is set up (a 200 from /api/me) we don't
 * re-poll within the session.
 */
export interface InstanceConfig {
  mode: 'standalone' | 'ha' | 'haos'
  instance_name: string
  /** Stable base32 id of this Social Home. Surfaced so cross-instance
   *  payloads (space-invite codes, the future GFS-redirect path) can
   *  embed it without an extra /api/friends round-trip. ``null`` when
   *  the federation tables aren't seeded yet (pre-setup cold start). */
  instance_id: string | null
  capabilities: string[]
  setup_required: boolean
  /** Vite content hash of the entry bundle the backend is currently
   *  serving (parsed from ``static/index.html``'s
   *  ``<script src="./assets/index-{hash}.js">`` tag). ``null`` when
   *  the backend isn't serving the SPA (Vite dev mode) or the bundle
   *  template is missing the canonical script tag. The
   *  ``SpaUpdateBanner`` compares this against the hash the running
   *  tab booted with to prompt the user when a deploy lands. */
  spa_bundle_hash?: string | null
}

export const instanceConfig = signal<InstanceConfig | null>(null)
export const instanceConfigError = signal<string | null>(null)

let inflight: Promise<InstanceConfig> | null = null

export async function loadInstanceConfig(): Promise<InstanceConfig> {
  if (inflight) return inflight
  inflight = api.get('/api/instance/config')
    .then((cfg) => {
      const typed = cfg as InstanceConfig
      instanceConfig.value = typed
      instanceConfigError.value = null
      return typed
    })
    .catch((err) => {
      instanceConfigError.value = err?.message || 'Failed to load instance config.'
      throw err
    })
    .finally(() => { inflight = null })
  return inflight
}

/** Same network call as :func:`loadInstanceConfig`, but does NOT use
 *  the inflight-dedup gate — a deliberate poll always issues a fresh
 *  request. Used by :class:`SpaUpdateBanner` so the bundle-hash check
 *  doesn't piggyback on a stale in-flight cold-start fetch. */
export async function recheckInstanceConfig(): Promise<InstanceConfig> {
  const cfg = await api.get('/api/instance/config') as InstanceConfig
  instanceConfig.value = cfg
  instanceConfigError.value = null
  return cfg
}

export function hasCapability(cap: string): boolean {
  return instanceConfig.value?.capabilities.includes(cap) ?? false
}

File: socialhome/routes/instance.py
"""Instance metadata route — what mode are we in, what can the SPA show?

`GET /api/instance/config` is the SPA's first call on cold start. The
response carries the deployment mode, the adapter's capability set, and
the first-boot setup flag so the SPA knows whether to route the user
to `/setup` instead of `/login`.

Public path — no auth required, intentionally. The SPA needs this
*before* it has a token.
"""

from __future__ import annotations

from aiohttp import web

from ..app_keys import (
    config_key,
    federation_repo_key,
    platform_adapter_key,
    setup_service_key,
)
from .base import BaseView
from .spa import _static_dir_key as _spa_static_dir_key
from .spa import get_spa_bundle_hash


class InstanceConfigView(BaseView):
    """``GET /api/instance/config`` — mode + capabilities + setup flag."""

    async def get(self) -> web.Response:
        config = self.svc(config_key)
        adapter = self.svc(platform_adapter_key)
        setup = self.svc(setup_service_key)
        capabilities = sorted(str(c) for c in adapter.capabilities)
        # SPA bundle hash — surfaced so a tab that's been open across a
        # backend deploy can poll this endpoint, notice the mismatch
        # against the hash it booted with, and prompt the user to
        # reload (see ``client/src/components/SpaUpdateBanner.tsx``).
        # ``None`` when the backend isn't serving the SPA (dev mode
        # behind ``pnpm dev``) or the bundle template is missing the
        # expected ``index-{hash}.js`` script tag.
        spa_bundle_hash: str | None = None
        static_dir = self.request.app.get(_spa_static_dir_key)
        if static_dir is not None:
            spa_bundle_hash = get_spa_bundle_hash(static_dir)
        # Surface this instance's stable id so SPA features that embed
        # the id in cross-instance payloads (e.g. space-invite codes)
        # don't have to round-trip through /api/friends to discover it.
        # Pre-setup the federation tables don't exist yet — degrade to
        # ``None`` rather than fail the cold-start config probe.
        instance_id: str | None = None
        fed_repo = self.request.app.get(federation_repo_key)
        if fed_repo is not None:
            try:
                identity = await fed_repo.get_local_identity()
                if identity is not None:
                    instance_id = identity.get("instance_id")
            except Exception:
                instance_id = None
        return web.json_response(
            {
                "mode": config.mode,
                "instance_name": config.instance_name,
                "instance_id": instance_id,
                "capabilities": capabilities,
                "setup_required": await setup.is_required(),
                "spa_bundle_hash": spa_bundle_hash,
            }
        )
